core/utils/messages.js

```diff
@@ -0,0 +1,215 @@
+
+const Msg = goog.require('Blockly.Msg');
+const stringUtils = goog.require('Blockly.utils.string');
+
+
+/**
+ * Parse a string with any number of interpolation tokens (%1, %2, ...).
+ * It will also replace string table references (e.g., %{bky_my_msg} and
+ * %{BKY_MY_MSG} will both be replaced with the value in
+ * Msg['MY_MSG']). Percentage sign characters '%' may be self-escaped
+ * (e.g., '%%').
+ * @param {string} message Text which might contain string table references and
+ *     interpolation tokens.
+ * @return {!Array<string|number>} Array of strings and numbers.
+ */
+const tokenizeInterpolation = function(message) {
+  return tokenizeInterpolation_(message, true);
+};
+exports.tokenizeInterpolation = tokenizeInterpolation;
+
+/**
+ * Replaces string table references in a message, if the message is a string.
+ * For example, "%{bky_my_msg}" and "%{BKY_MY_MSG}" will both be replaced with
+ * the value in Msg['MY_MSG'].
+ * @param {string|?} message Message, which may be a string that contains
+ *     string table references.
+ * @return {string} String with message references replaced.
+ */
+const replaceReferences = function(message) {
+  if (typeof message != 'string') {
+    return message;
+  }
+  const interpolatedResult = tokenizeInterpolation_(message, false);
+  // When parseInterpolationTokens == false, interpolatedResult should be at
+  // most length 1.
+  return interpolatedResult.length ? String(interpolatedResult[0]) : '';
+};
+exports.replaceReferences = replaceReferences;
+
+/**
+ * Validates that any %{MSG_KEY} references in the message refer to keys of
+ * the Msg string table.
+ * @param {string} message Text which might contain string table references.
+ * @return {boolean} True if all message references have matching values.
+ *     Otherwise, false.
+ */
+const checkReferences = function(message) {
+  let validSoFar = true;
+
+  const msgTable = Msg;
+
+  // TODO (#1169): Implement support for other string tables,
+  // prefixes other than BKY_.
+  const m = message.match(/%{BKY_[A-Z]\w*}/ig);
+  for (let i = 0; i < m.length; i++) {
+    const msgKey = m[i].toUpperCase();
+    if (msgTable[msgKey.slice(6, -1)] == undefined) {
+      console.warn('No message string for ' + m[i] + ' in ' + message);
+      validSoFar = false;  // Continue to report other errors.
+    }
+  }
+
+  return validSoFar;
+};
+exports.checkReferences = checkReferences;
+
+/**
+ * Internal implementation of the message reference and interpolation token
+ * parsing used by tokenizeInterpolation() and replaceMessageReferences().
+ * @param {string} message Text which might contain string table references and
+ *     interpolation tokens.
+ * @param {boolean} parseInterpolationTokens Option to parse numeric
+ *     interpolation tokens (%1, %2, ...) when true.
+ * @return {!Array<string|number>} Array of strings and numbers.
+ */
+const tokenizeInterpolation_ = function(message, parseInterpolationTokens) {
```

Comment:
This is not marked private, but it has the underscore. The old function was also not marked private, but I don't have context on if that's intentional and this is another `bindEvent_` scenario where we made this available for external folks.

I think either we:
a) Intentionally make this public, remove the underscore (and rename it to avoid clashing), and add a deprecation notice to the old function (and add it to renames file)
b) make this private and keep the underscore

depending on the intent with this function. I don't have context on that though, maybe Rachel?

Reply:
I think it's private by default because it's not exported. How would you feel about renaming it to `tokenizeInterpolationInternal` but otherwise leaving as is?

Reply:
Oh, duh, forgot to check if it's exported. But good call on renaming, since I think we moved away from underscores on non-exported private methods. Thanks!